packages/transformers/graphql/src/GraphQLTransformer.js

```diff
@@ -1,57 +1,32 @@
 // @flow
 import {Transformer} from '@parcel/plugin';
 
-const IMPORT_RE = /^# *import +['"](.*)['"] *;? *$/;
-const NEWLINE = /\r\n|[\n\r\u2028\u2029]/;
+import {parse, print, Source} from 'graphql/language';
+import {stripIgnoredCharacters} from 'graphql/utilities';
+import {processDocumentImports} from 'graphql-import-macro';
 
 export default new Transformer({
   async transform({asset, options, resolve}) {
-    // Peer dependency of graphql-tag
-    await options.packageManager.resolve('graphql', asset.filePath);
-
-    let gql = await options.packageManager.require(
-      'graphql-tag',
-      asset.filePath,
-      {autoinstall: options.autoinstall},
-    );
-
-    let gqlMap = new Map();
-    const traverseImports = (name: string, assetCode: string) => {
-      gqlMap.set(name, assetCode);
-
-      return Promise.all(
-        assetCode
-          .split(NEWLINE)
-          .map(line => line.match(IMPORT_RE))
-          .filter(match => !!match)
-          // $FlowFixMe
-          .map(async ([, importName]) => {
-            let resolved = await resolve(name, importName);
-
-            if (gqlMap.has(resolved)) {
-              return;
-            }
-
-            let code = await options.inputFS.readFile(resolved, 'utf8');
-            asset.addIncludedFile({
-              filePath: resolved,
-            });
-
-            await traverseImports(resolved, code);
-          }),
+    const document = parse(new Source(await asset.getCode(), asset.filePath));
+
+    const expandedDocument = await processDocumentImports(document, loadImport);
+
+    async function loadImport(to, from) {
+      const filePath = await resolve(to, from);
+
+      asset.addIncludedFile({filePath});
+
+      return parse(
+        new Source(await options.inputFS.readFile(filePath, 'utf-8'), filePath),
       );
-    };
+    }
 
-    await traverseImports(asset.filePath, await asset.getCode());
+    const generated = asset.env.minify
+      ? stripIgnoredCharacters(print(expandedDocument))
+      : print(expandedDocument);
 
     asset.type = 'js';
-    asset.setCode(
-      `module.exports=${JSON.stringify(
-        gql([...gqlMap.values()].join('\n')),
-        null,
-        2,
-      )};`,
-    );
+    asset.setCode(`module.exports=${JSON.stringify(generated)};`);
```

Comment:
What did the old transformer's export return and what does your version return? A string or an object representing the AST?

Reply:
My mistake! The old code does return an AST, which is also the behaviour in Parcel@1:
```
      `module.exports=${JSON.stringify(	
        gql([...gqlMap.values()].join('\n')),	
        null,	
        2,	
      )};`
```
This was not obvious to me in the above code. Using `gql` vs `parse` and the `join`.

I can make a separate issue to further discuss this as it is a breaking change. I feel like `v1` -> `v2` is the perfect time to make this breaking change.

---

I feel like a string is a much more usable return type:
-   It is ready to send to the server, so that the user does not need to have the `graphql` package on the frontend to format the value before use. Is the AST is desired it can be parsed.
-   It is surprising as other bundlers return a string. I was not expecting an AST the first time I used parcel.
-   It is significantly smaller.
    ```
    query GetUser($userId: ID!) {user(id: $userId) {id name isViewerFriend}}
    ```
    vs
    ```
    {
    "kind": "Document",
    "definitions": [
     {
      "kind": "OperationDefinition",
      "operation": "query",
      "name": {
        "kind": "Name",
        "value": "GetUser",
        "loc": {
          "start": 6,
          "end": 13
        }
      },
      "variableDefinitions": [
        {
          "kind": "VariableDefinition",
          "variable": {
            "kind": "Variable",
            "name": {
              "kind": "Name",
              "value": "userId",
              "loc": {
                "start": 15,
                "end": 21
              }
            },
            "loc": {
              "start": 14,
              "end": 21
            }
          },
          "type": {
            "kind": "NonNullType",
            "type": {
              "kind": "NamedType",
              "name": {
                "kind": "Name",
                "value": "ID",
                "loc": {
                  "start": 23,
                  "end": 25
                }
              },
              "loc": {
                "start": 23,
                "end": 25
              }
            },
            "loc": {
              "start": 23,
              "end": 26
            }
          },
          "directives": [],
          "loc": {
            "start": 14,
            "end": 26
          }
        }
      ],
      "directives": [],
      "selectionSet": {
        "kind": "SelectionSet",
        "selections": [
          {
            "kind": "Field",
            "name": {
              "kind": "Name",
              "value": "user",
              "loc": {
                "start": 29,
                "end": 33
              }
            },
            "arguments": [
              {
                "kind": "Argument",
                "name": {
                  "kind": "Name",
                  "value": "id",
                  "loc": {
                    "start": 34,
                    "end": 36
                  }
                },
                "value": {
                  "kind": "Variable",
                  "name": {
                    "kind": "Name",
                    "value": "userId",
                    "loc": {
                      "start": 39,
                      "end": 45
                    }
                  },
                  "loc": {
                    "start": 38,
                    "end": 45
                  }
                },
                "loc": {
                  "start": 34,
                  "end": 45
                }
              }
            ],
            "directives": [],
            "selectionSet": {
              "kind": "SelectionSet",
              "selections": [
                {
                  "kind": "Field",
                  "name": {
                    "kind": "Name",
                    "value": "id",
                    "loc": {
                      "start": 48,
                      "end": 50
                    }
                  },
                  "arguments": [],
                  "directives": [],
                  "loc": {
                    "start": 48,
                    "end": 50
                  }
                },
                {
                  "kind": "Field",
                  "name": {
                    "kind": "Name",
                    "value": "name",
                    "loc": {
                      "start": 51,
                      "end": 55
                    }
                  },
                  "arguments": [],
                  "directives": [],
                  "loc": {
                    "start": 51,
                    "end": 55
                  }
                },
                {
                  "kind": "Field",
                  "name": {
                    "kind": "Name",
                    "value": "isViewerFriend",
                    "loc": {
                      "start": 56,
                      "end": 70
                    }
                  },
                  "arguments": [],
                  "directives": [],
                  "loc": {
                    "start": 56,
                    "end": 70
                  }
                }
              ],
              "loc": {
                "start": 47,
                "end": 71
              }
            },
            "loc": {
              "start": 29,
              "end": 71
            }
          }
        ],
        "loc": {
          "start": 28,
          "end": 72
        }
      },
      "loc": {
        "start": 0,
        "end": 72
      }
     }
    ],
    "loc": {
    "start": 0,
    "end": 72
    }
    }
    ```

Reply:
> I can make a separate issue to further discuss this as it is a breaking change. I feel like v1 -> v2 is the perfect time to make this breaking change.

Indeed, the question at this point is not whether it's a breaking change but rather what is more useful when using GraphQL/more versatile.

I have no graphql experience, but: if other tools also emit a string and the most common client libraries work on a string schema, we should return a string.

Reply:
From https://graphql.org/code/#javascript-1:
String:
-   https://github.com/prisma-labs/graphql-request
-   https://github.com/kadirahq/lokka
-   

String -> custom
-   https://github.com/yoshuawuyts/nanographql

Other
-   https://github.com/Houfeng/gq-loader (webpack loader that exports an async function)
-   https://docs.amplify.aws (cannot find the graphql client docs)
-   https://github.com/grafoojs/grafoo/tree/master/packages/core (babel plugin)

Reply:
I made a separate issue to track the breaking change: https://github.com/parcel-bundler/parcel/issues/4511